sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -628,6 +576,73 @@ class GpuDynamicPartitionDataSingleWriter(
     }
   }
 
+  /**
+   * Write a Batch.
+   *
+   * Note: The `batch` will be closed in this function.
+   *
+   * @param batch the batch to be written
+   * @param maxRecordsPerFile the max number of rows per file
+   * @param partPath the partition directory
+   */
+  private def writeBatch(batch: ColumnarBatch, maxRecordsPerFile: Long, partPath: String) = {
+    val (table, dataTypes) = withResource(batch) { _ =>
+      (GpuColumnVector.from(batch), GpuColumnVector.extractTypes(batch))
+    }
+    writeTable(table, dataTypes, maxRecordsPerFile, partPath)
+  }
+
+
+  /**
+   * Write a Table.
+   *
+   * Note: The `table` will be closed in this function.
+   *
+   * @param table the table to be written
+   * @param outDataTypes the data types of the table
+   * @param maxRecordsPerFile the max number of rows per file
+   * @param partPath the partition directory
+   */
+  private def writeTable(
+    table: Table, outDataTypes: Array[DataType], maxRecordsPerFile: Long, partPath: String) = {
+    val splitIndexes = getSplitIndexes(
+      maxRecordsPerFile,
+      currentWriterStatus.recordsInFile,
+      table.getRowCount()
+    )
+    var needNewWriter = currentWriterStatus.recordsInFile >= maxRecordsPerFile
+
+    val tabs = withResource(table) { _ => table.contiguousSplit(splitIndexes: _*)}
+    withResource(tabs) { _ =>
+      tabs.foreach(b => {
+        if (needNewWriter) {
+          currentWriterStatus.fileCounter += 1
+          assert(currentWriterStatus.fileCounter <= MAX_FILE_COUNTER,
+            s"File counter ${currentWriterStatus.fileCounter} " +
+              s"is beyond max value $MAX_FILE_COUNTER")
+
+          // will create a new file, close the old writer
+          if (currentWriterStatus != null) {
+            releaseWriter(currentWriterStatus.outputWriter)
+          }
+
+          // create a new writer and update the writer in the status
+          currentWriterStatus.outputWriter =
+            newWriter(partPath, None, currentWriterStatus.fileCounter)
+          currentWriterStatus.recordsInFile = 0
+        }
+        withResource(b.getTable()) {tab =>
+          val bc = GpuColumnVector.from(tab, outDataTypes)
```

Comment:
it doesn't look like we are closing this (`cb`)

Reply:
This is not a leak because `currentWriterStatus.outputWriter.write` closes the batch.

Reply:
ah ok, thanks. It would be nice if `write` was called `writeAndClose` as we've done in other places in the code. But that is not a must fix.